app/mailers/roster_mailer.rb

```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+class RosterMailer < ApplicationMailer
+  before_action { @roster = params[:roster] }
+
+  def open_dates_alert
+    @open_dates = @roster.check_for_open_dates_between(Time.zone.now, 2.weeks.from_now)
+    mail to: @roster.admins.pluck(:email), subject: "Upcoming dates are uncovered for #{@roster.name} On-Call"
+    mail.perform_deliveries = false if @open_dates.empty?
```

Comment:
I'm not sure I'm really on board with having the `perform_deliveries = false` logic in the mailer itself here. To me it doesn't make sense that a mailer should sometimes "cancel itself" - the responsibilities of how/whether or not to send an email are usually delegated to the callee (i.e. usually callee decides whether to `deliver_now`, `deliver_later`, or not send at all).

Keeping in the vein of my review, such responsibility would fall on the `Job` (i.e. _it_ would gather the `open_dates` data for each roster, determine whether or not to send an email for, then just forward those dates here for formatting).

Reply:
I don't disagree, so I'll make the change. With that said, `mail.perform_deliveries = false` is one of the `after_action` examples in the [Action Mailer Rails guide](https://guides.rubyonrails.org/action_mailer_basics.html#action-mailer-callbacks) .

Reply:
Interesting - I gave it a quick readthrough but missed that section.